hip-0000-hedera-services-calendar-versioning.md

```diff
@@ -0,0 +1,224 @@
+- As an engineering manager, project manager, key stakeholder, or governing council member, I want to understand what versioned release is being discussed or communicated at any given time, and when such release will be deployed to various pre-production and production environments, and when the version will be replaced by a future version.
+- As a developer, tester, researcher, or user conducting historical analysis of previous Node Software releases using production network state data produced by such releases, I want to understand what version of software produced a given state and what version(s) can be utilized to load such state into a test environment
+- As a developer or tester, I want to be able to invoke certain upgrade / migration pathways in a testing environment without the invocation actually performing any operations that may significantly upgrade or otherwise alter an existing state produced by another hashgraph network.
+  
+## Specification
+
+### Specification Overview
+
+The Node Software project, collectively developed in the [hashgraph/hedera-services](https://github.com/hashgraph/hedera-services) GitHub repository, will adjust its versioning, release tagging, release planning, and communications regarding new releases to utilize the CalVer scheme modeled under the [Ubuntu Release Cycle](https://ubuntu.com/about/release-cycle).
+
+While the Node Software project releases under a different cadence from that of the Ubuntu project, the versioning scheme is compatible with any calendar-based software release cycle. Since the scheme works for any calendar-based cycle, it can also continue to work for the Node Software project should the project ever decide to adjust its release cadence in the future.
+
+In the process of adopting CalVer / Ubuntu version scheme, the project will preserve and maintain the current existing principles, processes, and tooling that have codified its use of SemVer up to this point. This includes but is not necessarily limited to:
+
+  - Documented guidelines and processes for creating and managing release branches and tags
+  - CI/CD and build pipelines that evaluate / compare version strings and numbers to determine a given course of action
+  - Code within the Node Software project itself that evaluates / compares version strings and numbers to determine a given course of action (though it is intended to enhance some of this functionality in the process of implementing the changes proposed in this HIP)
+
+### Scheme
+
+**Short Version:** Node Software project versioning will follow the convention `YY.MM.z<-modifier>`, while also providing 2 optional Build Number settings to allow invocation of extremely minor version increments to support certain cases that only affect development and testing of the Node Software.
```

Comment:
Based on format notations of [CalVer](https://calver.org/), shouldn't it be `YYYY-0M-0D-z<-modifier>`?
YYYY: 4 digit year (Let's not introduce the possibility of 2K bug again)
0M: Zero padded month (01, 02, 03,...,11, 12)
0D: Zero padded day (01, 02, 03, ....30,31)
z: Patch number (starts at 0)
modifier: `alpha | beta | rc1`

Reply:
You can go either way on the Year part with CalVer

> As seen in the [case studies](https://calver.org/#case_studies) below, projects have found more than one useful way to leverage dates in their versions. Rather than choose a single scheme, CalVer introduces standard terminology for developers, in addition to the "semantic" versions:
>
>     YYYY - Full year - 2006, 2016, 2106
>     YY - Short year - 6, 16, 106
>     0Y - Zero-padded year - 06, 16, 106
>     MM - Short month - 1, 2 ... 11, 12
>     0M - Zero-padded month - 01, 02 ... 11, 12
>     WW - Short week (since start of year) - 1, 2, 33, 52
>     0W - Zero-padded week - 01, 02, 33, 52
>     DD - Short day - 1, 2 ... 30, 31
>     0D - Zero-padded day - 01, 02 ... 30, 31

So you can pick and choose the specific format that works best for your project and its community.


> 4 digit year (Let's not introduce the possibility of 2K bug again)
As shown above, you can go either way on this. I am suggesting we use the 2 digit year (at least for now) for the following reasons:

1. It's shorter, simpler, and easier to communicate
2. It's good enough for Ubuntu, one of the most popular Linux distributions in the world
3. A Y2K-style bug is not a concern for roughly ~70 years. By then, Ubuntu and other projects using the 2-digit format will have to come to a resolution
4. I don't think choosing 2-digit today necessarily discounts the possibility of switching to 4-digit at such point it becomes more desirable or necessary. At such point we could potentially go from a `99.0.0` release to the next release being `2100.0.0`. So long as the next major version number is bigger than the last, the rules should still apply and the versions should be evaluated properly. It is possible there could be gaps in some SemVer implementations that do not properly handle 4 digits in the major versions, but if those gaps do exist I would think they need to be fixed eventually

TL;DR I don't think 2-digit year puts us in a corner today and we can still go to 4-digit year later. But both options are valid under CalVer scheme

Reply:
Quick follow up: Further discussions around this particular topic have surfaced that the year 2100+ is already accounted for in the CalVer spec: both the "short year" `YY` format and "zero-padded year" `0Y` format examples demonstrate these 2-digit formats can increase to 3 digits when the need arises:

> YY - Short year - 6, 16, 106
> 0Y - Zero-padded year - 06, 16, 106

So if we chose the 2-digit year format today as proposed, the first release for the software in the year 2100 would either look like `100.1.0` (using short-year format) or `100.01.0` (using zero-padded-year format)

With that understanding of what the spec provides out of the box for us, I believe the tl;dr is now along these lines:
1. CalVer already solved the year 2100+ problem should a project choose to use either 2-digit year format it provides
2. Additionally, another option still exists where the 4-year format could be adopted in the future if desired.
3. In addition to both options 1 and 2, a 3rd option exists where the project itself could be renamed / started fresh in some way - start the year 2100 by taking the end result of version `99.12.x`, start a new repo with a refreshed name, and start the versioning back at `00.1.x`, archiving the previous project for posterity

The 3 options are ranked in level of disruption: option 1 should provide little-to-no disruption to the project. Option 2 could require some version comparison logic or dependent libraries to be updated to support a 4-digit number in the "MAJOR" position. Option 3 is highly disruptive, but depending on where the project goes that far into the future, circumstances around or even before the year 2100 could make such a route desirable. There is no way to say from where we stand today.

I think this gives us enough answer to this question to be ok to proceed and mark this piece of feedback resolved. Thanks for raising the feedback though, it sparked some very good discussions and thinking around this particular problem!